// @okxweb3/coin-solana/dist/sdk/spl/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.programSupportsExtensions = exports.NATIVE_MINT_2022 = exports.NATIVE_MINT = exports.ASSOCIATED_TOKEN_PROGRAM_ID = exports.TOKEN_2022_PROGRAM_ID = exports.TOKEN_PROGRAM_ID = void 0;
const web3_1 = require("../web3");
exports.TOKEN_PROGRAM_ID = new web3_1.PublicKey('TokenkegQfeZyiNwAJbNbGKPFXCWuBvf9Ss623VQ5DA');
exports.TOKEN_2022_PROGRAM_ID = new web3_1.PublicKey('TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb');
exports.ASSOCIATED_TOKEN_PROGRAM_ID = new web3_1.PublicKey('ATokenGPvbdGVxr1b2hvZbsiqW5xWH25efTNsLJA8knL');
exports.NATIVE_MINT = new web3_1.PublicKey('So11111111111111111111111111111111111111112');
exports.NATIVE_MINT_2022 = new web3_1.PublicKey('9pan9bMn5HatX4EJdBwg9VgCa7Uz5HL8N1m5D3NdXejP');
function programSupportsExtensions(programId) {
    if (programId === exports.TOKEN_PROGRAM_ID) {
        return false;
    }
    else {
        return true;
    }
}
exports.programSupportsExtensions = programSupportsExtensions;
