// @okxweb3/coin-cosmos/dist/amino/signDoc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.serializeSignDoc = exports.makeSignDoc = exports.sortedJsonStringify = void 0;
const crypto_lib_1 = require("@okxweb3/crypto-lib");
function sortedObject(obj) {
    if (typeof obj !== "object" || obj === null) {
        return obj;
    }
    if (Array.isArray(obj)) {
        return obj.map(sortedObject);
    }
    const sortedKeys = Object.keys(obj).sort();
    const result = {};
    sortedKeys.forEach((key) => {
        result[key] = sortedObject(obj[key]);
    });
    return result;
}
function sortedJsonStringify(obj) {
    return JSON.stringify(sortedObject(obj));
}
exports.sortedJsonStringify = sortedJsonStringify;
function makeSignDoc(msgs, fee, chainId, memo, accountNumber, sequence) {
    return {
        chain_id: chainId,
        account_number: crypto_lib_1.math.Uint53.fromString(accountNumber.toString()).toString(),
        sequence: crypto_lib_1.math.Uint53.fromString(sequence.toString()).toString(),
        fee: fee,
        msgs: msgs,
        memo: memo || "",
    };
}
exports.makeSignDoc = makeSignDoc;
function serializeSignDoc(signDoc) {
    return crypto_lib_1.base.toUtf8(sortedJsonStringify(signDoc));
}
exports.serializeSignDoc = serializeSignDoc;
