// @okxweb3/coin-solana/dist/sdk/metaplex/plugins/nftModule/DelegateInput.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseTokenMetadataDelegateInput = void 0;
const types_1 = require("../../types");
const pdas_1 = require("./pdas");
const parseTokenMetadataDelegateInput = (mint, input, programs) => {
    if ('updateAuthority' in input) {
        return {
            isTokenDelegate: false,
            delegate: input.delegate,
            approver: input.updateAuthority,
            delegateRecord: (0, pdas_1.getMetadataDelegateRecord)({
                mint,
                type: input.type,
                updateAuthority: input.updateAuthority,
                delegate: (0, types_1.isSigner)(input.delegate)
                    ? input.delegate.publicKey
                    : input.delegate,
                programs,
            }),
        };
    }
    const tokenAccount = input.token ??
        (0, pdas_1.getAssociatedTokenAccount)({
            mint,
            owner: input.owner,
            programs,
        });
    return {
        isTokenDelegate: true,
        delegate: input.delegate,
        approver: input.owner,
        delegateRecord: (0, pdas_1.getTokenRecord)({
            mint,
            token: tokenAccount,
            programs,
        }),
        tokenAccount,
    };
};
exports.parseTokenMetadataDelegateInput = parseTokenMetadataDelegateInput;
